OpenSim/Common/Path.cpp

```diff
@@ -0,0 +1,223 @@
+// CONSTRUCTORS
+//=============================================================================
+//_____________________________________________________________________________
+
+//_____________________________________________________________________________
+/**
+ * Constructors
+ */
+Path::Path(const std::string path,
+           const char separator,
+           const std::string invalidChars) :
+    _separator(separator), _invalidChars(invalidChars), _isAbsolute(false)
+{
+    // check if path is empty
+    if (path.empty()) return;
+
+    // check if this is an absolute path
+    char firstChar = path.at(0);
+    if (firstChar == _separator) _isAbsolute = true;
+    
+    // separate the path (there may be multiple separators at the front)
+    size_t start = path.find_first_not_of(separator);
+    while (start != std::string::npos) {
+        size_t end = path.find_first_of(separator, start);
+        // if last segment (without a "/" at the end)
+        if (end == std::string::npos) {
+            end = path.find_last_not_of(separator, std::string::npos) + 1;
+        }
+        std::string pathElement = path.substr(start, end - start);
+        appendPathElement(pathElement);
+        start = path.find_first_not_of(separator, end + 1);
+    }
+
+    cleanPath();
+}
+
+Path::Path(std::vector<std::string> pathVec,
+           const char separator,
+           const std::string invalidChars,
+           bool isAbsolute) :
+    _path(pathVec), _separator(separator), _invalidChars(invalidChars),
+    _isAbsolute(isAbsolute)
+{
+    if (_path.empty()) return;
+    cleanPath();
+}
+
+std::string Path::getString() const
+{
+    std::string outString;
+    if (_isAbsolute) outString.append(1, _separator);
+    for (size_t i = 0; i < getPathLength(); ++i) {
+        outString.append(_path[i]);
+        if (i != getPathLength() - 1) {
+            outString.append(1, _separator);
+        }
+    }
+
+    return outString;
+}
+
+void Path::insertPathElement(size_t pos, const std::string& pathElement) 
+{
+    if (pos > getPathLength()) {
+        OPENSIM_THROW(Exception, "insertPathElement: pos must be between 0 and path length");
+    }
+
+    if (!isLegalPathElement(pathElement)) {
+        OPENSIM_THROW(Exception, "invalid character used in path element");
+    }
+    _path.insert(_path.begin() + pos, pathElement);
+}
+
+void Path::erasePathElement(size_t pos)
+{
+    if (pos > getPathLength() - 1) {
+        OPENSIM_THROW(Exception, "erasePathElement: pos is out of range of elements");
+    }
+
+    _path.erase(_path.begin() + pos);
+}
+
+std::vector<std::string> Path::formAbsolutePathVec(Path* otherPath) 
+{
+    if (_isAbsolute) {
+        return _path;
+    }
+    
+    if (!otherPath->_isAbsolute) {
+        OPENSIM_THROW(Exception, "getOatherPathVec: otherPath must be an absolute path");
+    }
+
+    std::vector<std::string> pathVec;
+    for (auto pathElement : otherPath->_path) {
```

Comment:
`auto pathElement` -> `const auto& pathElement`. And similarly for other places where this comes up.